File: crates/store/src/footprint.rs
//! What a Job's worktree held when the Job stopped.
//!
//! # A record, not a reading
//!
//! Everywhere else a footprint appears it is read live: `fleet`'s watcher opens
//! the worktree while a Drone works, and `get_diff` opens it again for a person
//! who asked. Both need the directory to still be there, and `armada clean`
//! gives worktrees back. **This is the one footprint that survives the
//! worktree**, because it was taken at the moment the Job stopped and written
//! down rather than re-derived afterwards.
//!
//! That is the whole reason it is not recomputed on read. A reading taken now
//! is a guess about a directory that may not exist; a row written at the
//! terminal transition is a record of one that did.
//!
//! # No drift mark on the row, and none derived here
//!
//! `outside_plan` is not a column, and [`crate::schema`]'s `V15` carries why.
//! What the record is measured against is [`crate::plan`]; the comparison is
//! made where the two are served and is stored in neither, for
//! [`crate::attempt`]'s reason.
//!
//! # The counts are here and on no live seam
//!
//! `added` and `deleted` arrived in [`V25`], because counting a file costs the
//! xdiff that renders its patch — 25ms over a hundred files, 90ms over four
//! hundred, against under a microsecond for the path list. That is affordable
//! once, on the transition that ends a Job, and it is not affordable on a
//! reading taken every two seconds inside a 250ms turn. So the record carries
//! them and `job.files_changed` does not.

use adapter_traits::{Change, ChangedFile, Counted, CountedFile, LineCount};
use core_model::{JobId, Timestamp};

use crate::error::{fault, LoadJobError, RowError, WriteError};
use crate::open::Store;
use crate::row::{column, string};

/// Version 25 — what each file in a footprint gained and lost.
///
/// Beside the table it changes, like [`V17`](crate::report::V17) and the
/// migrations after it: `schema.rs` is at the 900 lines the gate refuses at.
///
/// **Nullable, and null is "not counted" rather than zero.** A binary file has
/// no patch to count and a rename that edited nothing is a real `0` — a column
/// that could not tell those apart would turn an unmeasured file into a file
/// that changed nothing.
///
/// **The pair may not half-arrive**, which is why the table is rebuilt rather
/// than altered: `SQLite` cannot add a `CHECK` to a table that exists, and the
/// counts come from one walk of one patch, so a row holding one number and not
/// the other is a state no reading can produce and none should be able to
/// write. **Nothing to backfill**: a footprint written before this has no
/// counts, which is what null says.
pub(crate) const V25: &str = r#"
CREATE TABLE job_footprint_files_counted (
    job_id  TEXT NOT NULL REFERENCES jobs(job_id),
    ordinal INTEGER NOT NULL,
    path    TEXT NOT NULL,
    change  TEXT NOT NULL,
    added   INTEGER,
    deleted INTEGER,
    PRIMARY KEY (job_id, ordinal),
    CHECK ((added IS NULL) = (deleted IS NULL))
) STRICT;

INSERT INTO job_footprint_files_counted (job_id, ordinal, path, change)
SELECT job_id, ordinal, path, change FROM job_footprint_files;

DROP TABLE job_footprint_files;

ALTER TABLE job_footprint_files_counted RENAME TO job_footprint_files;
"#;

/// One Job's footprint as it was written down, and when.
///
/// **Absent and empty are different sentences.** No [`Footprinted`] at all is a
/// Job nothing recorded — still running, or finished before the record existed.
/// One with no files is a worktree that was opened and held no change, which is
/// what a `diff_nonempty` check refuses.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Footprinted {
    /// The reading, in the order it was found. Never re-sorted.
    ///
    /// A file with no [`LineCount`] is one nothing could count, which is not a
    /// file that changed no lines.
    pub files: Vec<CountedFile>,
    /// When the reading was taken, which is when the Job reached its terminal
    /// status and not when anybody asked for it.
    pub recorded_at: Timestamp,
}

impl Store {
    /// Write down what the worktree held, replacing whatever a previous write
    /// left.
    ///
    /// **Replacing, because a Job reaches a terminal status once and a second
    /// write means the first was wrong** — a Fleet that recorded, crashed
    /// before the transition landed and recorded again on the way back through
    /// must end with one reading rather than two interleaved.
    ///
    /// A transaction, because the header and its rows are one record: a header
    /// with nobody's rows under it says the Job changed nothing, which is a
    /// claim no torn write may make.
    pub fn record_footprint(
        &mut self,
        job_id: &JobId,
        changed: &Counted,
        at: &Timestamp,
    ) -> Result<(), WriteError> {
        let tx = self
            .conn
            .transaction()
            .map_err(fault("starting the footprint record"))
            .map_err(WriteError::Database)?;

        tx.execute(
            "DELETE FROM job_footprint_files WHERE job_id = ?1",
            rusqlite::params![job_id.as_str()],
        )
        .map_err(fault("clearing an earlier footprint"))
        .map_err(WriteError::Database)?;

        tx.execute(
            "INSERT INTO job_footprint (job_id, recorded_at) VALUES (?1, ?2)
             ON CONFLICT (job_id) DO UPDATE SET recorded_at = excluded.recorded_at",
            rusqlite::params![job_id.as_str(), at.as_str()],
        )
        .map_err(fault("writing a job's footprint"))
        .map_err(WriteError::Database)?;

        for (ordinal, file) in changed.files().iter().enumerate() {
            tx.execute(
                "INSERT INTO job_footprint_files (job_id, ordinal, path, change, added, deleted)
                 VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
                rusqlite::params![
                    job_id.as_str(),
                    ordinal as i64,
                    file.path(),
                    file.change().as_wire(),
                    file.lines().map(|lines| i64::from(lines.added())),
                    file.lines().map(|lines| i64::from(lines.deleted())),
                ],
            )
            .map_err(fault("writing a changed file"))
            .map_err(WriteError::Database)?;
        }

        tx.commit()
            .map_err(fault("committing the footprint record"))
            .map_err(WriteError::Database)
    }

    /// What was written down for this Job, or `None` where nothing was.
    ///
    /// **The header decides**, and the files are read only where there is one.
    /// Asking the file rows alone could not tell a Job that was never recorded
    /// from one recorded as having changed nothing.
    ///
    /// A `change` value this build does not know is
    /// [`RowError::UnknownEnumValue`] and refuses the whole read, rather than
    /// dropping the file: a footprint short one row reads as work that was
    /// never done, which is the failure mode this record exists to end.
    pub fn footprint(&self, job_id: &JobId) -> Result<Option<Footprinted>, LoadJobError> {
        let recorded: Option<String> = self
            .conn
            .query_row(
                "SELECT recorded_at FROM job_footprint WHERE job_id = ?1",
                rusqlite::params![job_id.as_str()],
                |row| row.get(0),
            )
            .map(Some)
            .or_else(|why| match why {
                rusqlite::Error::QueryReturnedNoRows => Ok(None),
                other => Err(RowError::Database(fault("reading a job's footprint")(
                    other,
                ))),
            })
            .map_err(LoadJobError::Unreadable)?;
        let Some(recorded_at) = recorded else {
            return Ok(None);
        };
        let files = self
            .collect(
                "SELECT path, change, added, deleted FROM job_footprint_files
                 WHERE job_id = ?1 ORDER BY ordinal",
                job_id,
                "reading a job's changed files",
                |row| {
                    let change = string(row, "change")?;
                    let file = ChangedFile::new(
                        string(row, "path")?,
                        Change::from_wire(&change).ok_or(RowError::UnknownEnumValue {
                            table: "job_footprint_files",
                            column: "change",
                            value: change,
                        })?,
                    );
                    Ok(CountedFile::new(file, counted(row)?))
                },
            )
            .map_err(LoadJobError::Unreadable)?;
        Ok(Some(Footprinted {
            files,
            recorded_at: Timestamp::from_rfc3339(recorded_at),
        }))
    }
}

/// What one stored row says the file gained and lost, or nothing where it was
/// never counted.
///
/// **A half-filled pair is a refusal, not an absence.** [`V25`]'s `CHECK` means
/// no write can make one, so a row holding one number and not the other was
/// written by something that is not this crate — and reading it as "not
/// counted" would let a file that lost four hundred lines report nothing at
/// all.
fn counted(row: &rusqlite::Row<'_>) -> Result<Option<LineCount>, RowError> {
    let added: Option<i64> = row
        .get("added")
        .map_err(column("job_footprint_files", "added"))?;
    let deleted: Option<i64> = row
        .get("deleted")
        .map_err(column("job_footprint_files", "deleted"))?;
    match (added, deleted) {
        (Some(added), Some(deleted)) => Ok(Some(LineCount::of(added as u32, deleted as u32))),
        (None, None) => Ok(None),
        _ => Err(RowError::MalformedColumn {
            table: "job_footprint_files",
            column: "added",
            detail: "one of the line counts is null and the other is not".to_string(),
        }),
    }
}

File: packages/protocol/src/footprint.ts
// What a finished job touched, as TypeScript sees it.
// `crates/ipc/src/work.rs`.
//
// **Its own file, because `protocol.ts` reached the 900 lines the gate refuses
// at.** These four types are one subject and the only one on that file with a
// natural seam: they are read together or not at all, nothing outside a
// finished job's record uses any of them, and they name no other wire type — so
// the import runs one way, from `protocol.ts` to here.
//
// **The record, not the live reading.** `ChangedFile` on a `job.files_changed`
// event is what a drone is doing now and stays in `protocol.ts` beside the
// event that carries it. This is what was written down when the job stopped,
// which is the one reading that survives `armada clean` giving the worktree
// back — and the only one carrying line counts, because counting costs the
// walk that renders the patch.

/**
 * What one job's worktree held when the job stopped. On `JobDetail` rather than
 * a read of its own: it is a path and a word per file, and Fleet asks for it
 * only where a job has one, so an open of a running job pays nothing.
 */
export type JobFootprint = {
  /** Every file, in the order the reading found them. */
  files: TouchedFile[];
  /**
   * When the reading was taken. **The instant the job stopped**, not the
   * instant it was asked for — which is what makes this a record and lets a
   * surface say so.
   */
  recorded_at: string;
  /**
   * What each run of each step said its work would be, in declaration order.
   * Since protocol 4.17, and absent from a fleet older than that.
   *
   * **Empty is the whole of "there is nothing to be outside of."** Every
   * `TouchedFile.planned_by` is then absent rather than empty, so a surface
   * that never reads this list still cannot draw an unmeasured path as one
   * that stayed in scope.
   */
  plans?: DeclaredPlan[];
};

/**
 * What one run of one step promised its work would be. Since protocol 4.17.
 *
 * **The promise, beside the record of what was done.** A footprint is the job's
 * whole work and a plan belongs to one step, so the two arrive side by side
 * rather than folded into one mark. A step that never declared has no entry: it
 * is silent, not counted.
 */
export type DeclaredPlan = {
  step_id: string;
  /**
   * Which run of that step declared it, one-based. **A step may be worked twice
   * and then declares twice**, and without this the two entries would read as
   * one step contradicting itself.
   */
  attempt: number;
  /** When the declaration was taken, by fleet's clock. */
  declared_at: string;
  /**
   * The paths the drone named, each covering everything beneath it at a segment
   * boundary. **Empty is a declaration of nothing**, which every changed path is
   * outside of — not a step that never declared.
   */
  paths: string[];
};

/**
 * One file a finished job touched.
 *
 * **Not `ChangedFile`, and the drift mark is the reason.** A live reading
 * carries `outside_plan` as a boolean, because the step being watched declared
 * the plan it is measured against. A record is the job's whole work, and the
 * step holding the pen when a job stops is usually not the step that scoped it
 * — so one boolean here could only be right by accident. `planned_by` names
 * steps rather than asserting a verdict.
 */
export type TouchedFile = {
  /** Repository-relative, exactly as git spells it. */
  path: string;
  /** The same closed set `ChangedFile.change` carries, left as `string`. */
  change: string;
  /**
   * The steps whose declared plan covers this path, in `JobFootprint.plans`
   * order. Since protocol 4.17.
   *
   * **Three readings, and the absent one is why this is not a boolean.** Absent
   * is a job where no step declared anything, so nothing was measured. Present
   * and empty is a path outside every plan that was declared — the drift a
   * finished job could not state before. Present with steps in it is a path one
   * of those steps promised.
   */
  planned_by?: string[];
  /**
   * What the file gained and lost. Since protocol 6.3.
   *
   * **Absent where nothing counted it** — a binary file, a footprint recorded
   * by a fleet older than this, or a patch the repository would not build.
   * Absent is not zero: a file moved without being edited is present and `0`.
   *
   * **`ChangedFile` has no counterpart and is not getting one.** The live
   * reading is taken every two seconds inside fleet's 250ms turn, and counting
   * is the same walk that renders the patch — 25ms over a hundred files, 90ms
   * over four hundred. This is read once, when the job stops.
   */
  lines?: LineCount;
};

/**
 * How many lines one file gained and lost. Since protocol 6.3.
 *
 * **One object rather than two optional numbers**, because the repository
 * answers both from one walk of one file's patch or answers neither.
 */
export type LineCount = {
  added: number;
  deleted: number;
};
